pylivetrader/backend/alpaca.py

```diff
@@ -549,50 +556,106 @@ def _symbol_bars(
         size:    str ('day', 'minute')
         _from:   str or pd.Timestamp
         to:      str or pd.Timestamp
         limit:   str or int
 
-        return: dict[str -> pd.DataFrame]
+        return: list[pd.DataFrame with columns MultiIndex [symbol -> OHLCV]]
         """
         assert size in ('day', 'minute')
 
-        if not (_from or to):
-            to = pd.to_datetime('now', utc=True).tz_convert('America/New_York')
-
-        if not (_from and to) and limit:
-            # temp workaround for less bars after masking by
-            # market hours
-            query_limit = limit
-            if query_limit is not None:
-                query_limit *= 2
-            if _from:
-                if size == 'day':
-                    to = _from + timedelta(days=query_limit+1)
-                else:
-                    to = _from + timedelta(minutes=query_limit+1)
-            else:
-                if size == 'day':
-                    _from = to - timedelta(days=query_limit+1)
-                else:
-                    _from = to - timedelta(minutes=query_limit+1)
+        assert (_from and to) or limit
+
+        if not (_from and to):
+            _from, to = self._get_from_and_to(size, limit, end_dt=to)
+
+        if self._use_polygon:
+            args = [[symbol, _from, to, size]for symbol in symbols]
+            return parallelize_with_multi_process(self._fetch)(args)
+        else:
+            # alpaca support get real-time data of multi stocks(<200) at once
+            splitlen = 199
+            parts = []
+            for i in range(0, len(symbols), splitlen):
+                part = symbols[i:i + splitlen]
+                parts.append(part)
+            args = [[part, _from, to, size] for part in parts]
+            return parallelize_with_multi_process(self._fetch, 10)(args)
+            # return parallelize(self._fetch)(args).values()
+
+    def _symbol_trades(self, symbols):
+        '''
+        Query last_trade in parallel for multiple symbols and
+        return in dict.
+
+        symbols: list[str]
+
+        return: dict[str -> polygon.Trade]
+        '''
 
         @skip_http_error((404, 504))
         def fetch(symbol):
             if self._use_polygon:
+                return self._api.polygon.last_trade(symbol)
+            else:
+                return self._api.get_last_trade(symbol)
+
+        return parallelize(fetch)(symbols)
+
+    def _get_from_and_to(self, size, limit, end_dt=None):
+        """
+        generate time interval for api
+
+        return:pd.Timestamp(tz=America/New_York)
+        """
+        if not end_dt:
+            end_dt = pd.to_datetime('now', utc=True).floor('min')
+        session_label = self._cal.minute_to_session_label(end_dt)
+        all_minutes: pd.DatetimeIndex = self._cal.all_minutes
+        all_sessions: pd.DatetimeIndex = self._cal.all_sessions
+        if size == 'minute':
+            if end_dt not in self._cal.minutes_for_session(session_label):
```

Comment:
document. why you do this. why is this valid. what is your logic

Reply:
end_dt may not be the trading time